zebrad/src/commands/start/sync.rs

```diff
@@ -198,100 +192,71 @@ where
             }
         }
 
-        self.request_blocks(download_set.into_iter().collect())
-            .await?;
+        self.prospective_tips
+            .retain(|tip| !download_set.contains(tip));
+
+        // ExtendTips Step 5
+        //
+        // Combine all elements of the remaining responses into a
+        // set, and queue download and verification of those blocks
+        self.request_blocks(
+            download_set
+                .into_iter()
+                .chain(self.prospective_tips.iter().cloned())
+                .collect(),
+        )
+        .await?;
 
         Ok(())
     }
 
     /// Queue downloads for each block that isn't currently known to our node
-    async fn request_blocks(&mut self, mut hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
-        hashes.retain(|hash| !self.known_block(hash));
-
+    async fn request_blocks(&mut self, hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
         for chunk in hashes.chunks(10usize) {
-            self.queue_download(chunk).await?;
-        }
-
-        Ok(())
-    }
-
-    /// Drive block downloading futures to completion and dispatch downloaded
-    /// blocks to the validator
-    async fn process_blocks(&mut self) -> Result<(), Report> {
-        info!(in_flight = self.block_requests.len(), "processing blocks");
-
-        while let Some(res) = self.block_requests.next().await {
-            match res.map_err::<Report, _>(|e| eyre!(e)) {
-                Ok(zn::Response::Blocks(blocks)) => {
-                    info!(count = blocks.len(), "received blocks");
-                    for block in blocks {
-                        let hash = block.as_ref().into();
-                        assert!(
-                            self.downloading.remove(&hash),
-                            "all received blocks should be explicitly requested and received once"
-                        );
-                        let _ = self.downloaded.insert(hash);
-                        self.validate_block(block).await?;
+            let set = chunk.iter().cloned().collect();
+
+            let request = self
+                .peer_set
+                .ready_and()
+                .await
+                .map_err(|e| eyre!(e))?
+                .call(zn::Request::BlocksByHash(set));
+
+            let mut state = self.state.clone();
+
+            let _ = tokio::spawn(async move {
+                match async move {
+                    let resp = request.await?;
+
+                    if let zn::Response::Blocks(blocks) = resp {
+                        debug!(count = blocks.len(), "received blocks");
+
+                        for block in blocks {
+                            state
+                                .ready_and()
+                                .await?
+                                .call(zs::Request::AddBlock { block })
+                                .await?;
```

Comment:
We need to verify the blocks before they're committed to the state, and the block verifications need to be driven on independent tasks.

They need to be on independent tasks because the verifications themselves will have inter-block dependencies (higher blocks may not be able to make progress before lower blocks, if they refer to data in the lower blocks), but processing blocks sequentially is a different dependency (blocks must be processed in the order they are received), and those dependencies might conflict, causing a deadlock.  To resolve/prevent this, each block verification needs to be on an independent task.

Reply:
Even worse, the checkpoint verifier makes all the blocks between a pair of checkpoints depend on each other.

In this case, lower blocks can't complete before higher blocks, because they depend on the chain back from the higher checkpoint. But higher blocks can't complete before lower blocks, because we want to add blocks to the state in order. So we wait for the whole set, verify the chain between checkpoints, and add them all in order.